src/biom_subsampled.cpp

```diff
@@ -0,0 +1,169 @@
+     const uint32_t cnt = obs_counts_resident[i];
+     _max_count = std::max(_max_count,cnt);
+     obs_data_resident[i] = (double**)malloc(sizeof(double*) * cnt);
+     if(obs_data_resident[i] == NULL) {
+        fprintf(stderr, "Failed to allocate %zd bytes; [%s]:%d\n", 
+                sizeof(double*) * cnt, __FILE__, __LINE__);
+        exit(EXIT_FAILURE);
+     }
+   }
+   max_count = _max_count;
+
+   for (uint32_t i=0; i<other.n_obs; i++) {
+     const uint32_t cnt = other.obs_counts_resident[i];
+     const uint32_t *idxs = other.obs_indices_resident[i];
+     double *data = other.obs_data_resident[i];
+     for (uint32_t j=0; j<cnt; j++) {
+       uint32_t local_i = idxs[j];
+       obs_data_resident[local_i][0] = &(data[j]);
+       obs_data_resident[local_i]++;
+     }
+   }
+
+   // we moved the pointers in place, rewind
+   for (uint32_t i=0; i<n_obs; i++) {
+      const uint32_t cnt = obs_counts_resident[i];
+      obs_data_resident[i] -=  cnt;
+   }
+}
+
+linked_sparse_transposed::~linked_sparse_transposed() {
+    if(obs_data_resident != NULL) {
+        free(obs_data_resident);
+        obs_data_resident = NULL;
+    }
+    if(obs_counts_resident != NULL) {
+       free(obs_counts_resident);
+        obs_counts_resident = NULL;
+    }
+}
+
+void linked_sparse_transposed::transposed_subsample_with_replacement(const uint32_t n, const uint32_t random_seed) {
+    std::default_random_engine generator(random_seed);
+
+    // use common buffer to minimize allocation costs
+    double *data_in = new double[max_count];  // original values
+    uint32_t *data_out = new uint32_t[max_count]; // computed values
+
+    for (uint32_t i=0; i<n_obs; i++) {
+        unsigned int length = obs_counts_resident[i];
+        double* *data_arr = obs_data_resident[i];
+
+        for (unsigned int j=0; j<length; j++) data_in[j] = *(data_arr[j]);
+        
+        std::discrete_distribution<uint32_t> multinomial(data_in, data_in+length);
+        for (unsigned int j=0; j<length; j++) data_out[j] = 0;
+        for (uint32_t j=0; j<n; j++) data_out[multinomial(generator)]++;
+
+        for (unsigned int j=0; j<length; j++) *(data_arr[j]) = data_out[j];
+    }
+    delete[] data_out;
+    delete[] data_in;
+}
+
+// =====================  sparse_data_subsampled  ==========================
+
+void sparse_data_subsampled::subsample_with_replacement(const uint32_t n, const uint32_t random_seed) {
+    linked_sparse_transposed transposed(*this);
+    transposed.transposed_subsample_with_replacement(n,random_seed);
+}
+
+// =====================  biom_subsampled  ==========================
+
+biom_subsampled::biom_subsampled(const biom_inmem &parent, const uint32_t n, const uint32_t random_seed) 
+  : biom_inmem(true)
+{
+   sparse_data_subsampled tmp_obj(parent.get_resident_obj(), true);
+   tmp_obj.subsample_with_replacement(n,random_seed);
+   copy_nonzero(parent,tmp_obj);
+
+    /* define a mapping between an ID and its corresponding offset */
```

Comment:
This block is replicated, would it make sense to have it as a protected member of a baseclass?

Reply:
I do not want to over-generalize yet. Focusing on correctness for now.